### PR TITLE
ADBDEV-3098-18: ORCA produces bogus plan for queries with CTE during handling distribution for Sequence children

### PR DESCRIPTION
**Pull request by RekGRpth**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp [hunk 1]
```diff
@@ -282,7 +282,8 @@ CPhysicalSequence::PdsRequired(CMemoryPool *mp,
 	}
 
 	// first child is non-singleton, request a non-singleton distribution on second child
-	return GPOS_NEW(mp) CDistributionSpecNonSingleton();
+	return GPOS_NEW(mp) CDistributionSpecNonSingleton(
+		false /* fAllowReplicated */, true /* fAllowEnforced */);
```
Comment (Stolb27) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp]: I think, this assumption will be incorrect, if CTE result is inserted to replicated table. There is no segment count shrinking in this case, and strict replicated distribution should be the best one.

File: src/backend/gporca/libgpopt/src/operators/CPhysical.cpp [hunk 2]
```diff
@@ -446,6 +447,13 @@ CPhysical::PdsUnary(CMemoryPool *mp, CExpressionHandle &exprhdl,
 		return pds;
 	}
 
+	if (CDistributionSpec::EdtNonSingleton == pdsRequired->Edt() &&
+		!CDistributionSpecNonSingleton::PdsConvert(pdsRequired)->FAllowReplicated())
+	{
+		pdsRequired->AddRef();
+		return pdsRequired;
+	}
```
Comment (Stolb27) [inline src/backend/gporca/libgpopt/src/operators/CPhysical.cpp:455]: Can we avoid appending Filter operator that useless at the first glance? Existing semantic of this node seems to be good for other purposes, as far as make it possible to stop enforcement propagation. But it's not a case for Sequence operator planning.

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CPhysical.cpp:455]: we may avoid appending filter https://github.com/arenadata/gpdb/pull/448

File: src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp (hunk 1 above)
Comment (Stolb27) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp]: As was shown by @RekGRpth, replicated distribution is bun unconditionally, an alternative with Hashed distribution will be chosen, and Broadcast above will be required.

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp]: rewrite conditions fixes insert into replicated table

**Pull request closed without merging**